File: src/main/java/com/github/lero4ka16/te4j/template/compiler/TemplateCompileProcess.java
/*
 *    Copyright 2021 Lero4ka16
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.github.lero4ka16.te4j.template.compiler;

import com.github.lero4ka16.te4j.Te4j;
import com.github.lero4ka16.te4j.include.IncludeTarget;
import com.github.lero4ka16.te4j.template.Template;
import com.github.lero4ka16.te4j.template.compiler.code.IterationCode;
import com.github.lero4ka16.te4j.template.compiler.code.RenderCode;
import com.github.lero4ka16.te4j.template.compiler.exp.ExpParsedList;
import com.github.lero4ka16.te4j.template.compiler.exp.ExpParser;
import com.github.lero4ka16.te4j.template.compiler.path.AbstractCompiledPath;
import com.github.lero4ka16.te4j.template.compiler.path.DefaultCompiledPath;
import com.github.lero4ka16.te4j.template.compiler.path.IncludeCompiledPath;
import com.github.lero4ka16.te4j.template.compiler.path.PathAccessor;
import com.github.lero4ka16.te4j.template.compiler.switchcase.SwitchCase;
import com.github.lero4ka16.te4j.template.context.TemplateContext;
import com.github.lero4ka16.te4j.template.environment.Environment;
import com.github.lero4ka16.te4j.template.environment.LoopEnvironment;
import com.github.lero4ka16.te4j.template.environment.PrimaryEnvironment;
import com.github.lero4ka16.te4j.template.exception.TemplateException;
import com.github.lero4ka16.te4j.template.method.TemplateMethodType;
import com.github.lero4ka16.te4j.template.method.impl.ConditionMethod;
import com.github.lero4ka16.te4j.template.method.impl.ForeachMethod;
import com.github.lero4ka16.te4j.template.method.impl.IncludeMethod;
import com.github.lero4ka16.te4j.template.method.impl.SwitchCaseMethod;
import com.github.lero4ka16.te4j.template.method.impl.ValueMethod;
import com.github.lero4ka16.te4j.template.output.TemplateOutput;
import com.github.lero4ka16.te4j.template.output.TemplateOutputBuffer;
import com.github.lero4ka16.te4j.template.output.TemplateOutputStream;
import com.github.lero4ka16.te4j.template.output.TemplateOutputString;
import com.github.lero4ka16.te4j.template.parse.ParsedTemplate;
import com.github.lero4ka16.te4j.template.path.TemplatePath;
import com.github.lero4ka16.te4j.template.path.TemplatePathIterator;
import com.github.lero4ka16.te4j.util.formatter.TextFormatter;
import com.github.lero4ka16.te4j.util.hash.Hash;
import com.github.lero4ka16.te4j.util.type.TypeInfo;
import com.github.lero4ka16.te4j.util.type.ref.ITypeRef;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.io.OutputStream;
import java.lang.reflect.Constructor;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

/**
 * @author lero4ka16
 */
public class TemplateCompileProcess<T> {

    private static final String OUTPUT_STREAM_CLASS
            = OutputStream.class.getName();

    private static final String TEMPLATE_OUTPUT_CLASS
            = TemplateOutput.class.getName();

    private static final String TEMPLATE_OUTPUT_STRING_CLASS
            = TemplateOutputString.class.getName();

    private static final String TEMPLATE_OUTPUT_STREAM_CLASS
            = TemplateOutputStream.class.getName();

    private static final String TEMPLATE_OUTPUT_BUFFER_CLASS
            = TemplateOutputBuffer.class.getName();

    private final ParsedTemplate template;
    private final TemplateContext context;

    private final ITypeRef<T> type;
    private final String parentFile;

    private final TemplateClassCompiler javaCompiler;

    private final Set<String> includes;

    private final PrimaryEnvironment primaryEnvironment;
    private final Map<String, Environment> environments = new HashMap<>();

    private final ExpParser expParser;

    private AtomicInteger nameCounter;

    private int outputType;
    private SwitchCase currentSwitchCase;

    public TemplateCompileProcess(@NotNull TemplateContext context, @NotNull ParsedTemplate template,
                                  @NotNull ITypeRef<T> type, @NotNull String parentFile) {
        this.type = type;
        this.parentFile = parentFile;

        this.primaryEnvironment = new PrimaryEnvironment("object", type.getType(), type.getRawType());

        this.context = context;

        this.template = template;
        this.expParser = new ExpParser(this::compilePathAccessor);

        this.javaCompiler = TemplateClassCompiler.current();
        this.javaCompiler.setTemplateType(type.getCanonicalName());

        this.includes = new HashSet<>();

        for (TemplatePath path : template.getPaths()) {
            if (path.getMethodType() == TemplateMethodType.INCLUDE) {
                IncludeMethod includeMethod = path.getMethod();

                if (includeMethod.getFile().hasValues()) {
                    continue;
                }

                includes.add(parentFile + "/" + includeMethod.getFile().format());
            }
        }
    }

    public Environment setEnvironment(String name, Environment environment) {
        return environments.put(name, environment);
    }

    public void addEnvironment(String name, Environment environment) {
        if (environments.containsKey(name)) {
            throw new TemplateException("Namespace already exists: " + name);
        }

        environments.put(name, environment);
    }

    public void removeEnvironment(String name) {
        environments.remove(name);
    }

    public List<AbstractCompiledPath> compilePaths(List<TemplatePath> paths) {
        return paths.stream().map(this::compilePaths).collect(Collectors.toList());
    }

    public AbstractCompiledPath compilePaths(TemplatePath path) {
        String id = "_" + nameCounter.incrementAndGet();

        TemplateMethodType methodType = path.getMethodType();

        switch (methodType) {
            case INCLUDE:
                return new IncludeCompiledPath(id, path.<IncludeMethod>getMethod().getFile(), path);
            case CASE:
                return new DefaultCompiledPath(id, compilePathAccessor(path.<SwitchCaseMethod>getMethod().getValue()), path);
            case VALUE:
            case CONDITION: {
                String value = path.getMethodType() == TemplateMethodType.VALUE
                        ? path.<ValueMethod>getMethod().getValue()
                        : path.<ConditionMethod>getMethod().getCondition();

                return new DefaultCompiledPath(id, expParser.recompile(value), path);
            }
            default:
                String value = path.<ForeachMethod>getMethod().getPath();

                return new DefaultCompiledPath(id, expParser.recompile(value), path);
        }
    }

    private PathAccessor compilePathAccessor(String path) {
        TemplatePathIterator iterator = new TemplatePathIterator(path);

        if (!iterator.hasNext()) {
            return null;
        }

        String element = iterator.next();

        if (!iterator.hasNext()) {
            switch (element) {
                case "true":
                    return PathAccessor.TRUE;
                case "false":
                    return PathAccessor.FALSE;
                case "null":
                    return PathAccessor.NULL;
            }
        }

        Environment environment = environments.get(element);

        if (environment == null) {
            iterator.previous();
            environment = primaryEnvironment;
        }

        return environment.resolve(iterator);
    }

    public void writePath(AbstractCompiledPath path, RenderCode out) {
        String accessorValue = path.getAccessorValue();

        switch (path.getMethodType()) {
            case VALUE:
                out.appendCodeSegment("out.put(" + accessorValue + ");");
                break;
            case CASE: {
                SwitchCaseMethod SwitchCaseMethod = path.getMethod();
                String value = SwitchCaseMethod.getValue();
                String from = SwitchCaseMethod.getFrom();

                TypeInfo info = path.getReturnType();

                Class<?> type;
                Object[] values;

                if (from == null) {
                    if (!info.isEnum()) {
                        throw new IllegalStateException(value + " is not enum");
                    }

                    type = (Class<?>) info.getType();
                    values = type.getEnumConstants();
                } else {
                    ExpParsedList parsedList = expParser.parseList(from);

                    values = parsedList.getValues();
                    type = parsedList.getType();
                }

                SwitchCase oldSwitchCase = currentSwitchCase;

                SwitchCase switchCase = new SwitchCase(type);
                currentSwitchCase = switchCase;

                out.append("switch (").append(accessorValue).append(") {");

                for (Object o : values) {
                    switchCase.setValue(o);
                    out.append("case ");

                    boolean string = o instanceof String;

                    if (string) out.append("\"");
                    out.append(o.toString());
                    if (string) out.append("\"");

                    out.append(": {");
                    out.appendTemplate(SwitchCaseMethod.getBlock());
                    out.append("break;}");
                }

                currentSwitchCase = oldSwitchCase;

                if (SwitchCaseMethod.getDefaultBlock() != null) {
                    out.append("default: {");
                    out.appendTemplate(SwitchCaseMethod.getDefaultBlock());
                    out.append("}");
                }

                out.append("}");
                break;
            }
            case INCLUDE: {
                IncludeMethod method = path.getMethod();

                String fileName;
                IncludeTarget file = method.getFile();

                if (file.hasValues()) {
                    SwitchCase switchCase = currentSwitchCase;

                    if (switchCase != null) {
                        fileName = file.format(switchCase.getValue());
                    } else {
                        fileName = file.format();
                    }

                    includes.add(parentFile + "/" + fileName);
                } else {
                    fileName = file.format();
                }

                ParsedTemplate template = context.parse(parentFile + "/" + fileName);

                if (template.hasPaths()) {
                    out.appendTemplate(template);
                } else {
                    out.appendTextSegment(template);
                }

                break;
            }
            case CONDITION: {
                ConditionMethod method = path.getMethod();
                ParsedTemplate block = method.getBlock();
                ParsedTemplate elseBlock = method.getElseBlock();

                out.append("if(").append(accessorValue).append("){");
                out.appendTemplate(block);

                if (elseBlock != null) {
                    out.append("} else {");
                    out.appendTemplate(elseBlock);
                }

                out.append("}");
                break;
            }
            case FOR: {
                ForeachMethod method = path.getMethod();

                TypeInfo returnType = path.getReturnType();
                Class<?> listType = returnType.getComponentType();

                if (listType == null) {
                    throw new IllegalStateException("No iterable type found for " + path.getAccessorValue());
                }

                ParsedTemplate template = method.getBlock();
                String as = method.getAs();

                String id = path.getId();
                String counter = "__counter_" + id + "_";

                LoopEnvironment loop = new LoopEnvironment(counter);

                IterationCode iterationCode = new IterationCode(
                        id, listType.getName(), accessorValue, counter, template, loop,
                        returnType.isArrayList(), returnType.isArray(),
                        !List.class.isAssignableFrom(returnType.getRawType())
                );

                Environment prevLoop = setEnvironment("loop", loop);
                addEnvironment(method.getAs(), new PrimaryEnvironment(iterationCode.getElementFieldName(), listType, listType));

                iterationCode.write(out);

                setEnvironment("loop", prevLoop);
                removeEnvironment(as);
                break;
            }
        }
    }

    private final Map<Hash, String> byteValues = new HashMap<>();

    public AtomicInteger getNameCounter() {
        return nameCounter;
    }

    public void addBytes(Integer field, byte[] bytes) {
        String fieldName = getOutputPrefix(outputType) + field;
        String prevFieldName = byteValues.put(Hash.forArray(bytes), fieldName);

        switch (outputType) {
            case Te4j.STRING:
                addContent("private final String " + fieldName + " = "
                        + (prevFieldName != null ? prevFieldName : getString(bytes)) + ";");
                break;
            case Te4j.BYTES:
                addContent("private final byte[] " + fieldName + " = "
                        + (prevFieldName != null ? prevFieldName : getString(bytes) + ".getBytes()") + ";");
                break;
        }
    }

    private String getString(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        sb.append('"');

        try {
            new TextFormatter(bytes)
                    .replace(context.getReplace())
                    .write(sb);
        } catch (IOException e) {
            e.printStackTrace();
        }

        sb.append("\"");

        return sb.toString();
    }

    public int getOutputType() {
        return outputType;
    }

    private void addRenderMethod(int outputType) {
        this.outputType = outputType;
        this.nameCounter = new AtomicInteger();

        StringBuilder sb = new StringBuilder();

        sb.append("private void render(").append(type.getCanonicalName()).append(" object, ");
        sb.append(outputType == Te4j.STRING
                ? TEMPLATE_OUTPUT_STRING_CLASS
                : TEMPLATE_OUTPUT_CLASS);
        sb.append(" out) {");

        RenderCode renderCode = new RenderCode(this);
        renderCode.appendTemplate(template);
        renderCode.flushTemplates();

        sb.append(renderCode);
        sb.append('}');

        byteValues.clear();

        addContent(sb.toString());
    }

    public void addRenderAsBytes(boolean hasBytesOptimization) {
        StringBuilder sb = new StringBuilder();
        sb.append("public byte[] renderAsBytes(").append(type.getCanonicalName()).append(" object) {");

        if (hasBytesOptimization) {
            sb.append(TEMPLATE_OUTPUT_BUFFER_CLASS).append(" result = bytesOptimized.get();");
        } else {
            sb.append(TEMPLATE_OUTPUT_STRING_CLASS).append(" result = stringOptimized.get();");
        }

        sb.append("result.reset();");
        sb.append("render(object, result);");
        sb.append("return result.toByteArray();");
        sb.append("}");

        addContent(sb.toString());
    }

    public void addRenderAsString(boolean hasStringOptimization) {
        StringBuilder sb = new StringBuilder();
        sb.append("public String renderAsString(").append(type.getCanonicalName()).append(" object) {");

        if (hasStringOptimization) {
            sb.append(TEMPLATE_OUTPUT_STRING_CLASS).append(" result = stringOptimized.get();");
        } else {
            sb.append(TEMPLATE_OUTPUT_BUFFER_CLASS).append(" result = bytesOptimized.get();");
        }

        sb.append("result.reset();");
        sb.append("render(object, result);");
        sb.append("return result.toString();");
        sb.append("}");

        addContent(sb.toString());
    }

    public void addRenderToStream(boolean hasBytesOptimization) {
        StringBuilder sb = new StringBuilder();

        sb.append("public void renderTo(").append(type.getCanonicalName()).append(" object, ");
        sb.append(OUTPUT_STREAM_CLASS).append(" os) throws java.io.IOException {");

        if (hasBytesOptimization) {
            sb.append("render(object, new ").append(TEMPLATE_OUTPUT_STREAM_CLASS).append("(os));");
        } else {
            sb.append("os.write(renderAsBytes(object));");
        }

        sb.append("}");

        addContent(sb.toString());
    }

    public void addContent(String src) {
        javaCompiler.addContent(src);
    }

    public @NotNull String getOutputPrefix(int bit) {
        switch (bit) {
            case 1:
                return "STRING_";
            case 2:
                return "BYTES_";
            default:
                throw new IllegalArgumentException("Undefined bit: " + bit);
        }
    }

    @SuppressWarnings("unchecked")
    public Template<T> compile() throws Exception {
        addEnvironment("this", primaryEnvironment);

        for (int outputType : Te4j.OUTPUT_TYPES) {
            if ((context.getOutputTypes() & outputType) == outputType) {
                addRenderMethod(outputType);
            }
        }

        addRenderAsString((context.getOutputTypes() & Te4j.STRING) != 0);
        addRenderAsBytes((context.getOutputTypes() & Te4j.BYTES) != 0);
        addRenderToStream((context.getOutputTypes() & Te4j.BYTES) != 0);

        StringBuilder includeBuilder = new StringBuilder();

        for (String include : includes) {
            if (includeBuilder.length() > 0) {
                includeBuilder.append(", ");
            }

            includeBuilder.append('"');
            includeBuilder.append(new TextFormatter(parentFile + "/" + include).format());
            includeBuilder.append('"');
        }

        addContent("private final String[] includes = new String[]{" + includeBuilder + "};");
        addContent("public String[] getIncludes() { return includes; }");

        Class<?> result = javaCompiler.compile();

        Constructor<?> constructor = result.getDeclaredConstructor();
        return (Template<T>) constructor.newInstance();
    }

}

File: src/main/kotlin/com/github/lero4ka16/te4j/template/compiler/TemplateCompileProcess.kt
/*
 *    Copyright 2021 Lero4ka16
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package com.github.lero4ka16.te4j.template.compiler

import com.github.lero4ka16.te4j.Te4j
import com.github.lero4ka16.te4j.expression.ExpList
import com.github.lero4ka16.te4j.expression.ExpParser
import com.github.lero4ka16.te4j.expression.ExpValue
import com.github.lero4ka16.te4j.include.IncludeTarget
import com.github.lero4ka16.te4j.template.Template
import com.github.lero4ka16.te4j.template.compiler.TemplateClassCompiler.Companion.current
import com.github.lero4ka16.te4j.template.compiler.code.IterationCode
import com.github.lero4ka16.te4j.template.compiler.code.RenderCode
import com.github.lero4ka16.te4j.template.compiler.path.AbstractCompiledPath
import com.github.lero4ka16.te4j.template.compiler.path.DefaultCompiledPath
import com.github.lero4ka16.te4j.template.compiler.path.IncludeCompiledPath
import com.github.lero4ka16.te4j.template.compiler.path.PathAccessor
import com.github.lero4ka16.te4j.template.compiler.switchcase.SwitchStatement
import com.github.lero4ka16.te4j.template.context.TemplateContext
import com.github.lero4ka16.te4j.template.environment.Environment
import com.github.lero4ka16.te4j.template.environment.LoopEnvironment
import com.github.lero4ka16.te4j.template.environment.PrimaryEnvironment
import com.github.lero4ka16.te4j.template.exception.TemplateException
import com.github.lero4ka16.te4j.template.method.TemplateMethodType
import com.github.lero4ka16.te4j.template.method.impl.Condition
import com.github.lero4ka16.te4j.template.method.impl.Foreach
import com.github.lero4ka16.te4j.template.method.impl.Include
import com.github.lero4ka16.te4j.template.method.impl.SwitchCase
import com.github.lero4ka16.te4j.template.method.impl.Value
import com.github.lero4ka16.te4j.template.output.TemplateOutput
import com.github.lero4ka16.te4j.template.output.TemplateOutputBuffer
import com.github.lero4ka16.te4j.template.output.TemplateOutputStream
import com.github.lero4ka16.te4j.template.output.TemplateOutputString
import com.github.lero4ka16.te4j.template.parse.ParsedTemplate
import com.github.lero4ka16.te4j.template.path.TemplatePath
import com.github.lero4ka16.te4j.template.path.TemplatePathIterator
import com.github.lero4ka16.te4j.util.formatter.TextFormatter
import com.github.lero4ka16.te4j.util.type.ref.ITypeRef
import java.io.OutputStream
import java.util.ArrayList
import java.util.concurrent.atomic.AtomicInteger
import java.util.stream.Collectors
import kotlin.collections.set

/**
 * @author lero4ka16
 */
class TemplateCompileProcess<BoundType>(
    val context: TemplateContext,
    private val template: ParsedTemplate,
    private val type: ITypeRef<BoundType>,
    private val parentFile: String
) {
    private val javaCompiler: TemplateClassCompiler
    private val includes: MutableSet<String>
    private val primaryEnvironment: PrimaryEnvironment = PrimaryEnvironment("object", type.type, type.rawType)
    private val environments: MutableMap<String, Environment?> = HashMap()
    private val expParser: ExpParser

    var nameCounter: AtomicInteger = AtomicInteger()

    var outputType = 0
        private set

    private var currentSwitchStatement: SwitchStatement? = null

    @Suppress("SameParameterValue")
    private fun setEnvironment(name: String, environment: Environment): Environment? {
        return environments.put(name, environment)
    }

    private fun addEnvironment(name: String, environment: Environment) {
        if (environments.containsKey(name)) {
            throw TemplateException("Namespace already exists: $name")
        }

        environments[name] = environment
    }

    private fun removeEnvironment(name: String) {
        environments.remove(name)
    }

    fun compilePaths(paths: List<TemplatePath>): List<AbstractCompiledPath> {
        return paths.stream().map { path: TemplatePath ->
            this.compilePaths(
                path
            )
        }.collect(Collectors.toList())
    }

    private fun compilePaths(path: TemplatePath): AbstractCompiledPath {
        val id = "_" + nameCounter.incrementAndGet()

        when (path.methodType) {
            TemplateMethodType.INCLUDE -> return IncludeCompiledPath(
                id,
                path,
                path.getMethod<Include>().file
            )
            TemplateMethodType.CASE -> return DefaultCompiledPath(
                id, path,
                compilePathAccessor(path.getMethod<SwitchCase>().value)!!
            )
            TemplateMethodType.VALUE, TemplateMethodType.CONDITION -> {
                val value =
                    if (path.methodType === TemplateMethodType.VALUE)
                        path.getMethod<Value>().value
                    else
                        path.getMethod<Condition>().condition
                val exp = expParser.parseExpression(value)
                val result = exp.compile()

                val newAccessor: PathAccessor = if (exp is ExpValue) {
                    val oldAccessor = exp.accessor
                    if (oldAccessor.accessor == result) {
                        oldAccessor
                    } else {
                        PathAccessor(exp.objectType, result)
                    }
                } else {
                    PathAccessor(exp.objectType, result)
                }

                return DefaultCompiledPath(id, path, newAccessor)
            }
            else -> {
                val value = path.getMethod<Foreach>().path
                val exp = expParser.parseExpression(value)
                val result = exp.compile()
                return DefaultCompiledPath(id, path, PathAccessor(exp.objectType, result))
            }
        }
    }

    private fun compilePathAccessor(path: String): PathAccessor? {
        val iterator = TemplatePathIterator(path)
        if (!iterator.hasNext()) {
            return null
        }
        val element = iterator.next()
        if (!iterator.hasNext()) {
            when (element) {
                "true" -> return PathAccessor.TRUE
                "false" -> return PathAccessor.FALSE
                "null" -> return PathAccessor.NULL
            }
        }
        var environment = environments[element]
        if (environment == null) {
            iterator.previous()
            environment = primaryEnvironment
        }
        return environment.resolve(iterator)
    }

    fun writePath(path: AbstractCompiledPath, out: RenderCode) {
        when (path.methodType) {
            TemplateMethodType.VALUE -> out.appendCodeSegment("out.put(" + path.accessorValue + ");")
            TemplateMethodType.CASE -> {
                val method: SwitchCase = path.getMethod()
                val value = method.value
                val from = method.from

                val info = path.returnType

                val type: Class<*>
                val values: Array<*>

                if (from == null) {
                    if (!info.isEnum) {
                        throw IllegalStateException("$value is not enum")
                    }

                    type = info.type as Class<*>
                    values = type.enumConstants
                } else {
                    val exp = expParser.parseExpression(from)

                    if (exp !is ExpList) {
                        throw UnsupportedOperationException()
                    }

                    values = exp.toArray()
                    type = exp.objectType.componentType!!
                }

                val oldSwitchStatement = currentSwitchStatement
                val switchStatement = SwitchStatement(type)

                currentSwitchStatement = switchStatement
                out.append("switch (").append(path.accessorValue).append(") {")
                for (o: Any? in values) {
                    switchStatement.value = o
                    out.append("case ")
                    val string = o is String
                    if (string) out.append("\"")
                    out.append(o.toString())
                    if (string) out.append("\"")
                    out.append(": {")
                    out.appendTemplate(method.block)
                    out.append("break;}")
                }
                currentSwitchStatement = oldSwitchStatement

                if (method.defaultBlock != null) {
                    out.append("default: {")
                    out.appendTemplate(method.defaultBlock)
                    out.append("}")
                }

                out.append("}")
            }
            TemplateMethodType.INCLUDE -> {
                val method: Include = path.getMethod()
                val fileName: String
                val file: IncludeTarget = method.file
                if (file.hasValues()) {
                    val switchStatement = currentSwitchStatement

                    fileName = if (switchStatement != null) {
                        file.format(switchStatement.value!!)
                    } else {
                        file.format()
                    }

                    includes.add("$parentFile/$fileName")
                } else {
                    fileName = file.format()
                }

                val template = context.parse("$parentFile/$fileName")

                if (template.hasPaths) {
                    out.appendTemplate(template)
                } else {
                    out.appendTextSegment(template)
                }
            }
            TemplateMethodType.CONDITION -> {
                val method = path.getMethod<Condition>()
                val block = method.block
                val elseBlock = method.elseBlock
                out.append("if(").append(path.accessorValue).append("){")
                out.appendTemplate(block)
                if (elseBlock != null) {
                    out.append("} else {")
                    out.appendTemplate(elseBlock)
                }
                out.append("}")
            }
            TemplateMethodType.FOR -> {
                val method = path.getMethod<Foreach>()

                val returnType = path.returnType

                val listType = returnType.componentType
                    ?: throw IllegalStateException("No iterable type found for " + path.accessorValue)

                val template = method.block

                val counterFieldName = "__counter_" + path.id
                val elementFieldName = "__element_" + path.id

                val loop = LoopEnvironment(counterFieldName)
                val prevLoop = setEnvironment("loop", loop)

                addEnvironment(method.element, PrimaryEnvironment(elementFieldName, listType, listType))

                val iterationCode = IterationCode(
                    listType.name, path.id, counterFieldName, elementFieldName,
                    path.accessorValue, template, returnType.isArrayList, returnType.isArray,
                    ArrayList::class.java.isAssignableFrom(returnType.rawType), loop
                )

                iterationCode.write(out)

                if (prevLoop != null) {
                    setEnvironment("loop", prevLoop)
                }

                removeEnvironment(method.element)
            }
        }
    }

    private val byteValues: MutableMap<String, String> = HashMap()

    fun getOutputPrefix(bit: Int): String {
        return when (bit) {
            Te4j.STRING -> "STRING_"
            Te4j.BYTES -> "BYTES_"
            else -> throw IllegalStateException("Unknown bit: $bit")
        }
    }

    fun addField(field: Int, value: String) {
        val fieldName = getOutputPrefix(outputType) + field
        val prevFieldName = byteValues.put(value, fieldName)

        when (outputType) {
            Te4j.STRING -> addContent("private final String $fieldName = ${prevFieldName ?: "\"$value\""};")
            Te4j.BYTES -> addContent("private final byte[] $fieldName = ${prevFieldName ?: "\"$value\".getBytes()"};")
        }
    }

    private fun addRenderMethod(outputType: Int) {
        this.nameCounter.set(0)
        this.outputType = outputType

        val sb = StringBuilder()
        sb.append("private void render(").append(type.canonicalName).append(" object, ")
        sb.append(if (outputType == Te4j.STRING) TEMPLATE_OUTPUT_STRING_CLASS else TEMPLATE_OUTPUT_CLASS)
        sb.append(" out) {")
        val renderCode = RenderCode(this)
        renderCode.appendTemplate(template)
        renderCode.flushTemplates()
        sb.append(renderCode)
        sb.append('}')
        byteValues.clear()
        addContent(sb.toString())
    }

    private fun addRenderAsBytes(hasBytesOptimization: Boolean) {
        val sb = StringBuilder()
        sb.append("public byte[] renderAsBytes(").append(type.canonicalName).append(" object) {")
        if (hasBytesOptimization) {
            sb.append(TEMPLATE_OUTPUT_BUFFER_CLASS).append(" result = bytesOptimized.get();")
        } else {
            sb.append(TEMPLATE_OUTPUT_STRING_CLASS).append(" result = stringOptimized.get();")
        }
        sb.append("result.reset();")
        sb.append("render(object, result);")
        sb.append("return result.toByteArray();")
        sb.append("}")
        addContent(sb.toString())
    }

    private fun addRenderAsString(hasStringOptimization: Boolean) {
        val sb = StringBuilder()
        sb.append("public String renderAsString(").append(type.canonicalName).append(" object) {")
        if (hasStringOptimization) {
            sb.append(TEMPLATE_OUTPUT_STRING_CLASS).append(" result = stringOptimized.get();")
        } else {
            sb.append(TEMPLATE_OUTPUT_BUFFER_CLASS).append(" result = bytesOptimized.get();")
        }
        sb.append("result.reset();")
        sb.append("render(object, result);")
        sb.append("return result.toString();")
        sb.append("}")
        addContent(sb.toString())
    }

    private fun addRenderToStream(hasBytesOptimization: Boolean) {
        val sb = StringBuilder()
        sb.append("public void renderTo(").append(type.canonicalName).append(" object, ")
        sb.append(OUTPUT_STREAM_CLASS).append(" os) throws java.io.IOException {")
        if (hasBytesOptimization) {
            sb.append("render(object, new ").append(TEMPLATE_OUTPUT_STREAM_CLASS).append("(os));")
        } else {
            sb.append("os.write(renderAsBytes(object));")
        }
        sb.append("}")
        addContent(sb.toString())
    }

    private fun addContent(src: String?) {
        javaCompiler += src
    }

    @Throws(Exception::class)
    fun compile(): Template<BoundType> {
        addEnvironment("this", primaryEnvironment)

        for (outputType: Int in Te4j.OUTPUT_TYPES) {
            if ((context.outputTypes and outputType) == outputType) {
                addRenderMethod(outputType)
            }
        }

        addRenderAsString((context.outputTypes and Te4j.STRING) != 0)
        addRenderAsBytes((context.outputTypes and Te4j.BYTES) != 0)
        addRenderToStream((context.outputTypes and Te4j.BYTES) != 0)

        val includeBuilder = StringBuilder()

        for (include: String in includes) {
            if (includeBuilder.isNotEmpty()) {
                includeBuilder.append(", ")
            }

            includeBuilder.append('"')
            includeBuilder.append(TextFormatter("$parentFile/$include").format())
            includeBuilder.append('"')
        }

        addContent("private final String[] includes = new String[]{$includeBuilder};")
        addContent("public String[] getIncludes() { return includes; }")

        val result = javaCompiler.compile()
        val constructor = result.getDeclaredConstructor()

        @Suppress("UNCHECKED_CAST")
        return constructor.newInstance() as Template<BoundType>
    }

    companion object {
        private val OUTPUT_STREAM_CLASS = OutputStream::class.java.name
        private val TEMPLATE_OUTPUT_CLASS = TemplateOutput::class.java.name
        private val TEMPLATE_OUTPUT_STRING_CLASS = TemplateOutputString::class.java.name
        private val TEMPLATE_OUTPUT_STREAM_CLASS = TemplateOutputStream::class.java.name
        private val TEMPLATE_OUTPUT_BUFFER_CLASS = TemplateOutputBuffer::class.java.name
    }

    init {
        expParser = ExpParser { path: String ->
            compilePathAccessor(
                path
            )
        }
        javaCompiler = current()
        javaCompiler.templateType = type.canonicalName

        includes = HashSet()
        for (path: TemplatePath in template.paths) {
            if (path.methodType === TemplateMethodType.INCLUDE) {
                val include: Include = path.getMethod()

                if (include.file.hasValues()) {
                    continue
                }

                includes.add(parentFile + "/" + include.file.format())
            }
        }
    }
}